### DISCUSSION TITLE
Choose files type/extension (mimetype)

### DISCUSSION THREAD

**Discussion by gauravkrp** [directus/directus#11733] [Feature Requests]
Currently Directus allows to upload any file type when selecting field type as file or files while creating. Only Images are specifically supported. I had multiple usecases where we needed to upload specific files such as pdf, csv or mp4 (videos).

*Feature Request* - Option to select what mimetypes a given file type field would allow. Either a multi select dropdown to allow for images, pdfs, videos or a string to pass option like `image/jpeg, application/pdf` etc

**Comment by joselcvarela** [comment]
> Only Images are specifically supported
What does this mean? We support preview of image, video and audio files.

In the meantime (I think this is already somewhere here), you can listen to action `files.upload` and throw an error if file is not what you are expecting. Although you also need to delete the file since all the upload needs to happen.
Also you can create a custom endpoint to achieve that, but I believe it's more complex.

**Reply by 3zzy** [reply]
Can't throw an error and delete the file within the same action I believe. 
Edited to add: Ah, turns out can't even show the error thrown from the action, must be in a filter which isn't available for files.
Is there a workaround?

**Reply by daawaan4x** [reply]
@3zzy copy/pasting the file interfaces into a separate package as an extension and then modifying the input element to filter specific file types only is probably the only way at this point

**Comment by benosman** [comment]
I also needed a similar restriction on filetype and looked at using the `files.upload` action.

It's a pity there isn't a files.upload filter where an error could be thrown before the upload takes place. I have a poc of this, but the issue is you then need to pass in and deal with a stream.

**Reply by LudovicSterlin** [reply]
Could we open a Feature Request on a filter event for `file.upload`. Is that already discussed somewhere ?

**Reply by rijkvanzanten** [reply]
Go for it! I didn't see any open feature requests on that as of right now 🙂

**Comment by dreerr** [comment]
Is there any news on that topic? The only way I see now is to do all of this with a seperate custom endpoint and upon successful validation (including busboy) passing it on to the directus files.

edit: I want to use a captcha for file upload including form fields.

**Comment by MentalGear** [comment]
Also, the mine type validation should be done on front and backend.

**Comment by piotr-cz** [comment]
Here is my workaround using [Custom API Hooks](https://docs.directus.io/extensions/hooks.html), given that I'd like to allow only `image/jpeg` type on the *image* field of *news* collection:

```ts
import { defineHook } from '@directus/extensions-sdk';
import type { FailedValidationException as TFailedValidationException } from '@directus/shared/exceptions';
import type { FilterHandler } from '@directus/shared/types';
import type { ItemsService } from 'directus/dist/services';
import type { File } from 'directus/dist/types';
// import { validatePayload } from '@directus/shared/utils';

export default defineHook(async ({ filter }, { exceptions, services, getSchema }) => {
	const { FailedValidationException }: { FailedValidationException: typeof TFailedValidationException } = exceptions;
	const { FilesService } = services;

	const filesService: ItemsService<File> = new FilesService({
		schema: await getSchema(),
	});

	const filterHandler: FilterHandler<{ image?: string | null }> = async (payload) => {
		if (payload.image) {
			const file = await filesService.readOne(payload.image);

			if (file?.type !== 'image/jpeg') {
				throw new FailedValidationException({
					message: '"image.type" must be [image/jpeg]',
					path: ['image', 'type'],
					type: 'any.only',
					context: {
						valids: ['image/jpeg'],
						label: 'image.type',
						value: file.type,
						key: 'type',
					},
				});
			}

			/*
			// Using directus validatePayload
			const validationErrors = validatePayload(
				{ "_and": [{"image": {"type": {"_eq": "image/jpeg"}}}] },
				{ image: file }
			);


			if (validationErrors[0] && validationErrors[0].details[0]) {
				throw new FailedValidationException(validationErrors[0].details[0]);
			}
			// */
		}

		return payload;
	};

	filter('news.items.create', filterHandler as FilterHandler);
	filter('news.items.update', filterHandler as FilterHandler);
});
```

**Reply by martinsotirov** [reply]
This doesn't work for me. I get `You don't have permission to access this.` when I try to get the file with `FilesService` in my extension.

**Reply by martinsotirov** [reply]
Never mind, it turned out only `fileService.readMany` has problems with permissions. When I iterate over the files of my gallery and get them one by one with `fileService.readOne`, it works.

**Comment by rijkvanzanten** [comment]
Heya! Thanks for opening this feature request!

This feature request has received over **15 votes** from the community. This means we'll move this feature request to the Under Review state! 

The Core team will schedule a meeting to review this request as soon as possible. The discussion will then be approved or denied. You may or may not be invited to join this meeting with the core team.

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._

**Reply by dbelyaeff** [reply]
Is there any progress? 10th of May was half a year ago :-)

**Reply by terrxo** [reply]
1 year as of now :)

**Reply by LasseRosenow** [reply]
This is not meant as an offense, but please don't ping this discussion without actually contributing new ideas or opinions. :)
I think the Directus Team has a lot of work and they will take care of this if they find time to do it.

Posting things like this post only annoys all the many people who are subscribed to discussion who then see that there is nothing new except someone asking if there is something new. Just be more patient please.

You maybe think that it is not a big problem, but there are many people always posting these  kind of questions everywhere and these days my github notifications are getting more and more flooded with posts like these. :)

So just a small reminder here :) Nothing personal, I think it is normal to feel frustrated :)

**Comment by AareFabrik** [comment]
Hello
Thank you for sharing your code @piotr-cz 
I adapted it to work with the new validation object and for audio files. Also I created a folder for only audio files and if the file was uploaded there and it is not audio it gets directly deleted. -> put your folder id there to make it work. 

```javascript
import { defineHook } from "@directus/extensions-sdk";
import { FailedValidationError } from "@directus/validation";

const audioTypes = [
  "audio/wav",
  "audio/webm",
  "audio/aac",
  "audio/ogg",
  "audio/mpeg",
];

export default defineHook(async ({ filter }, { services, getSchema }) => {
  const { FilesService } = services;

  const filesService = new FilesService({
    schema: await getSchema(),
  });

  // filter("items.create", async (input, { collection }) => {
  const filterHandler = async (payload) => {

    if (payload.audio) {
      const file = await filesService.readOne(payload.audio);

      if (!audioTypes.includes(file.type)) {
        if (file.folder === "830d2c33-95c3-4b2e-b677-a0b1e1cfee5a") {
          await filesService.deleteOne(payload.audio);
        }
        throw new FailedValidationError({
          field: "audio",
          type: "eq",
          valid: audioTypes,
          substring: "one of these types: [" + audioTypes.join(", ") + "]",
        });
      }
    }
  };
  filter("posts.items.create", filterHandler);
  filter("posts.items.update", filterHandler);
});
```

**Comment by dithom** [comment]
@AareFabrik Great solution! However, there currently is an issue with custom validation messages in the build, just as an info. https://github.com/directus/directus/issues/24861

**Comment by iocouto** [comment]
I've just gone down the rabbit hole looking for a solution for this. I just wanted to document that at this point the known workarounds are to either use the permissions system to create a rule to allow uploading only a certain file type (which then applies to the entire Directus instance), or by trying to create a hook/automation that informs the user the file type is not allowed, and then deletes the file *after* the file has been uploaded (not a very elegant solution, which really feels like a hack). 

What is required in most instances is a **field-specific, mime-based validation** that applies primarily to the frontend: when the user goes to select files to upload — or select existing files from the file library — they should only be able to select files with the specific mime type(s) allowed, and the others should be greyed out. 

I hope this makes it clearer why this is still a necessity — I have been unable to find any existing extensions that can help with this use-case...

**Reply by JamesW1** [reply]
Thanks @iocouto - we're tracking this one on our roadmap and hope to get to it soon. If you register your interest here, you'll be notified when there is an update:
https://roadmap.directus.io/roadmap?id=1846ab1f-72d3-46a8-adb9-5d83ade6466b

I'll also be sure to circle back to this thread when there's progress to share.

**Comment by pmorin17** [comment]
This feature is really a must for us, our users are not all tech-savvy and might try to upload a video file in a field dedicated to pdf files.
As you can understand, this ends up being an important issue.

Mimetype validation is great, but we also want to have some filesize validation.

I ended up making a single flow that we can reuse for specific fields in different collections. It is far from ideal, but it gets the job done.

1. Event blocking flow on items.create and items.update for the collections we want.
2. Run Script, depending on what collection the flow was triggered from, we check for specific fields that were created/updated. We then build an array of file IDs and a file type. You could also put a specific filesize limit here on each specific field.
3. Read Data, we fetch all the files from our array of IDs
4. Run Script, we check if each file has a valid mimetype depending on the filetype previously mentionned. We also validate the file size here, we have an array with the maximum filesize for each mimetype we allow. We also build our custom error message here depending on the error (file type or file size) with the file name so that the user knows what's wrong.
5. Throw Error, we show the custom error built in the previous operation.

I don't feel comfortable sharing the whole flow, it's quite messy, but that's pretty much the whole logic.

I didn't find a way to delete invalid files without causing user experience issues.

**Comment by Tommy1223454** [comment]
Makes sense, but I’d be careful not to stop at “allowed mimetypes”.

That’s useful, but it’s still not the same as inspecting whether the uploaded bytes actually match the claimed type, or whether the file should be trusted enough to persist.

I built a Node.js OSS package for that deeper inspection layer:
https://github.com/pompelmi/pompelmi

So I tend to think of MIME allowlists as one layer, and server-side inspection as the next one.
